File: 2023/day1/day1.py
import re
class Day1:
    def __init__(self):
        self.lines = open("input.txt",'r').read().split("\n")
        self.calibration_sum = 0

    def first_puzzle(self):
        numbers = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

        for line in self.lines:
            scan_res = [m.group(1) for m in re.finditer(r"(?=(\d|{}))".format("|".join(numbers)), line)]
            digits = self.process_numbers(scan_res)
            calibration_value = int(f"{digits[0]}{digits[-1]}")
            self.calibration_sum += calibration_value

        print(self.calibration_sum)

    def process_numbers(self, digits):
        numbers = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
        processed_digits = []

        for d in digits:
            if d in numbers:
                processed_digits.append(numbers.index(d) + 1)
            else:
                processed_digits.append(d)

        return processed_digits


day1 = Day1()
day1.first_puzzle()

File: 2023/day1/day1.rb
require 'benchmark'
class Day1

    attr_accessor :lines, :calibration_sum, :numbers
    
    
    def initialize
        self.lines = File.read("input.txt").split("\n")
        self.calibration_sum = 0
    end

    def first_puzzle
        lines.each do |l|
            scan_res = l.scan(/(?=(\d))/).flatten
            calibration_value =  "#{scan_res.first}#{scan_res.last}".to_i
            self.calibration_sum += calibration_value
        end;
        puts calibration_sum
    end

    def second_puzzle
        lines.each do |l|
            scan_res = l.scan(/(?=(\d|one|two|three|four|five|six|seven|eight|nine))/).flatten
            calibration_value =  "#{process_numbers(scan_res.first)}#{process_numbers(scan_res.last)}".to_i
            self.calibration_sum += calibration_value
        end;
        puts calibration_sum
    end

    def process_numbers(d)
        n = ["one","two","three","four","five","six","seven","eight","nine"]
        if n.include?(d)
            return n.find_index(d)+1
        else
            return d
        end
    end
end

duration = Benchmark.realtime do 
    Day1.new.first_puzzle
    Day1.new.second_puzzle
end
puts duration